src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs

```diff
@@ -553,11 +553,24 @@ protected override void Dispose(bool disposing)
             {
                 if (_disposed)
                     return;
-                if (disposing && _producerTask.Status == TaskStatus.Running)
+
+                if (disposing)
                 {
-                    _toProduce.Post(0);
+                    _toProduce.Complete();
                     _producerTask.Wait();
+
+                    // Complete the consumer after the producerTask has finished, since producerTask could
+                    // have posted more items to _toConsume.
+                    _toConsume.Complete();
+
+                    // Drain both BufferBlocks - this prevents what appears to be memory leaks when using the VS Debugger
+                    // because if a BufferBlock still contains items, its underlying Tasks are not getting completed.
+                    // See https://github.com/dotnet/corefx/issues/30582 for the VS Debugger issue.
+                    // See also https://github.com/dotnet/machinelearning/issues/4399
+                    _toProduce.TryReceiveAll(out _);
+                    _toConsume.TryReceiveAll(out _);
```

Comment:
In what situation does the LoopProducerWorker exit without consuming everything?  I see the `if (numRows < requested)` condition, but I'm wondering if that's required, or if removing that and these TryReceiveAlls would be better.

Reply:
The `_toConsume` can easily get in a situation where everything isn't consumed. All that would need to happen is for the caller to create one of these Cursors, enumerate over a few rows and then exit early without reading all the way to the end of the cursor. In that situation, `_toConsume` can still have items ready to consume, but no one got them.

The `_toProduce` gets into this situation exactly where you pointed out - the `if (numRows < requested)` it exits early and doesn't drain the `_toProduce` buffer.

We could change LoopProducerWorker to not exit early and always consume everything in `_toProduce`, but we would still need to leave `_toConsume.TryReceiveAll` here.

Thoughts?

Reply:
What you have is fine.

(As an aside, I expect if you were to switch to using Channel instead of BufferBlock, you wouldn't see this debugger-induced leak, even without consuming everything from the block.)

Reply:
Ok, I will leave this as-is for now. But I've logged https://github.com/dotnet/machinelearning/issues/4482 to replace the BufferBlock usage with Channel.